### DISCUSSION TITLE
How to debug PHP code?

### DISCUSSION THREAD

**Discussion by PetersOtto** [datenstrom/community#824] [Ask a question] [answered]
Hello!
I am interested to know how do you develop extensions. Is there a workflow I should use?

I am not a professional coder. I learned all myself with tutorials from the internet. So, please be lenient with me :)

At the moment I am trying modify the `image.php`. I want to have two image sizes, one for mobile and one for the rest.

At the moment I write my code into the `image.php`. Then I try to load an image into the Yellow-Editor (drag and drop) and see if it work or not. I can't see intermediate steps. Something like `echo $fileName` would be great, to see what is insert in this variable.

I am using Ubuntu 20.04 as Desktop with LAMP where Yellow is installed on. I am using VS Code as Code-Editor.

Best regards :)

**Comment by PetersOtto** [comment]
Hello.
By using the `system/extensions/yellow-website.log` with `$this->yellow->log("info", "Debug Otto '" . strtok($variablename, " ") . "'");` and  activate the `CoreDebugMode` in the `system/extensions/yellow-system.ini` it was me possible to find bugs in my code.

**Comment by markseuffert** [comment, marked as answer]
Here are some starting points for debugging source code:

* Use a debugger. You are probably used to work with a debugger if you are coming from a programming language like C and C++. You can set break points in your source code, the software stops when it reaches a break point and shows the current variables. You can step trough you source code, line by line and see what happens. It's super comfortable, once you have a debugger set up in your favourite development environment. Have a look at this [introduction video for Xdebug](https://www.youtube.com/watch?v=SlluEqL0oOI). 

* Use `echo` or [similar PHP functions](https://stackoverflow.com/questions/1647322/whats-the-difference-between-echo-print-and-print-r-in-php). We all do it from time to time. You can show variables on screen and/or the current web page. It's super quick and a decent tool for every day work. Sometimes I prefer echos over a debugger, because you can format the debug output and make it human friendly. When you are working on a new algorithm or with a lot of input data, it allows you to quickly spot anomalies.

* Use `$this->yellow->toolbox->log()` from the [API for developers](https://datenstrom.se/yellow/help/api-for-developers). You can write the debug output to a log file. Sometimes I do this, because the debug output can interfere with the normal output. When you're working close to protocol level, for example in the middle of generating HTTP response headers, it's better to keep the debug output separate.

* Activate the debug mode in Datenstrom Yellow. You can use the debug mode to investigate the cause of a problem in more detail, to show the stack trace of a program or if you are curious about how Datenstrom Yellow works. To activate the debug mode on your website open file `system/extensions/yellow-system.ini` and change `CoreDebugMode: 1`. Depending on the debug level, more or less information are shown on screen. [Learn more about troubleshooting](https://datenstrom.se/yellow/help/troubleshooting).

Hope this helps and happy coding.

**Reply by PetersOtto** [reply]
Thank you for the tips.
